addon/record-data.js

```diff
@@ -622,6 +621,7 @@ export default class FragmentRecordData extends RecordData {
 
   willCommit() {
     for (const [key, behavior] of Object.entries(this._fragmentBehavior)) {
+      this._initializeFragmentDefaultValue(key);
```

Comment:
ember-data's Snapshot constructor calls .get on every model attribute, which has the side effect of initializing fragments' default values. Ember-data v4 changed when it creates the snapshot.

In ember-data 3.28, hooks are called in this order:

    model.save -> snapshots created -> recordData.willCommit -> recordData.didCommit

And in ember-data 4+

    model.save -> recordData.willCommit -> snapshots created -> recordData.didCommit

To replicate the problem:
1. Construct a model containing a fragment with a default value
2. Immediately call model.save
3. RecordData.willCommit will ignore the fragment because it doesn't exist
4. ember-data's Snapshot will initialize the fragment in uncommitted state
5. RecordData.didCommit will fail because the fragment is not in-flight:

  > Error: Attempted to handle event `didCommit` on <name:null> while in state root.loaded.created.uncommitted. 

This line ensures that all default values are initialized by willCommit.

Reply:
4598753d0499d18f36aedde72889ff8a3c968fd7 makes RecordData responsible for default values instead of the attribute. We no longer have to invoke the attribute getter for its side effect and we no longer depend on the Snapshot constructor.